### ISSUE TITLE
math: explore using constants and specialized functions to speed up functions

### ISSUE THREAD

**Issue by ericlagergren** [ericlagergren/decimal#60]
Let's consider `Exp`.

`Exp` uses the argument reduction `exp(z) = 10**k * exp(r) where r + k*log(10)`. We find this by:

1. Set `k = log10(e) * z`
2. Set `r = x - k*log(10)`
3. Return `10**k * e**r

This requires a fast way to compute `log10(e)`, since computing `log10(e) = log(e) / log(10)` is much too slow, as well as a fast way to compute `log(10)`.

If we have N-precision constants, our crossover for speed (the point where we become faster than other libraries, in this case libmpdec) for `Exp` is around 1000 digits of precision. With N=5000, we achieve about a 12% speedup until 1500 digits of precision where that nearly doubles to 23%. However, once we hit N digits of precision, our speedup drops down to 2% (our bounds for the constant are `[0, N)`). At 7500 we're down to a measly 0.45%. At 15000 we're about 55x _slower_.

We have a specialized routine for `log(10)` which uses a continued fraction representation of `log(10)` that jumps straight into the CF sequence, skipping range reduction and the general overhead of the `Log` function. Unfortunately, all we have for `log10(e)` is calling `Log10` with the result of a specialized function for `e`.

To recap: we need some way to quickly compute `log10(e)` and `log(10)`. One way is to use constants, but that limits us to `[0, N)` digits of precision. Another is to use constants and then set up the specialized functions to _start_ where the constants leave off. However, the downside is we'll have some _very_ large decimals hanging around. We could create these on demand, but we'd still have some _very_ large constant strings hanging around.
